File: src/main/java/org/stt/command/Activities.java
package org.stt.command;

import net.engio.mbassy.bus.BusRuntime;
import net.engio.mbassy.bus.IMessagePublication;
import net.engio.mbassy.bus.MBassador;
import net.engio.mbassy.bus.common.PubSubSupport;
import org.stt.model.ItemDeleted;
import org.stt.model.ItemInserted;
import org.stt.model.ItemReplaced;
import org.stt.model.TimeTrackingItem;
import org.stt.persistence.ItemPersister;
import org.stt.query.Criteria;
import org.stt.query.TimeTrackingItemQueries;
import org.stt.time.DateTimes;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.Collection;
import java.util.Optional;

import static java.util.Objects.requireNonNull;

/**
 * Command "side" of tracking activities
 */
@Singleton
public class Activities implements CommandHandler {
    private ItemPersister persister;
    private final TimeTrackingItemQueries queries;
    private final PubSubSupport<Object> publisher;

    @Inject
    public Activities(ItemPersister persister,
                      TimeTrackingItemQueries queries,
                      Optional<MBassador<Object>> publisher) {
        this.publisher = requireNonNull(publisher).map(PubSubSupport.class::cast)
                .orElseGet(DoNotPublish::new);
        this.persister = requireNonNull(persister);
        this.queries = requireNonNull(queries);
    }

    @Override
    public void addNewActivity(NewActivity command) {
        requireNonNull(command);
        TimeTrackingItem newItem = command.newItem;
        Optional<TimeTrackingItem> potentialItemToReplace = ongoingItemThatWouldEnd(newItem);
        if (!potentialItemToReplace.isPresent()) {
            potentialItemToReplace = itemWithEditedActivity(newItem);
        }

        if (potentialItemToReplace.isPresent()) {
            TimeTrackingItem itemToReplace = potentialItemToReplace.get();
            persister.replace(itemToReplace, command.newItem);
            publisher.publish(new ItemReplaced(itemToReplace, command.newItem));
        } else {
            persister.persist(command.newItem);
            publisher.publish(new ItemInserted(command.newItem));
        }
    }

    private Optional<TimeTrackingItem> itemWithEditedActivity(TimeTrackingItem newItem) {
        Criteria criteria = new Criteria()
                .withStartsAt(newItem.getStart());
        newItem.getEnd().ifPresent(criteria::withEndsAt);
        return queries.queryItems(criteria).findAny();
    }

    private Optional<TimeTrackingItem> ongoingItemThatWouldEnd(TimeTrackingItem newItem) {
        return queries.getLastItem()
                .filter(timeTrackingItem ->
                        !timeTrackingItem.getEnd().isPresent()
                                && newItem.getStart().equals(timeTrackingItem.getStart()));
    }

    @Override
    public void endCurrentActivity(EndCurrentItem command) {
        requireNonNull(command);
        queries.getOngoingItem()
                .ifPresent(item -> {
                    TimeTrackingItem derivedItem = item.withEnd(command.endAt);
                    persister.replace(item, derivedItem);
                    publisher.publish(new ItemReplaced(item, derivedItem));
                });
    }

    @Override
    public void removeActivity(RemoveActivity command) {
        requireNonNull(command);
        persister.delete(command.itemToDelete);
        publisher.publish(new ItemDeleted(command.itemToDelete));
    }

    @Override
    public void removeActivityAndCloseGap(RemoveActivity command) {
        requireNonNull(command);
        TimeTrackingItemQueries.AdjacentItems adjacentItems = queries.getAdjacentItems(command.itemToDelete);
        Optional<TimeTrackingItem> previous = adjacentItems.previousItem();
        Optional<TimeTrackingItem> next = adjacentItems.nextItem();

        if (previous.isPresent()) {
            TimeTrackingItem previousItem = previous.get();
            if (next.isPresent() && previousAndNextActivitiesMatch(previousItem, next.get())) {
                TimeTrackingItem replaceAllWith = next.get().getEnd()
                        .map(previousItem::withEnd).orElse(previousItem.withPendingEnd());
                persister.persist(replaceAllWith);
                publisher.publish(new ItemInserted(replaceAllWith));
                return;
            }
            if (DateTimes.isOnSameDay(previousItem.getStart(), command.itemToDelete.getStart())
                    && !command.itemToDelete.getEnd().isPresent()) {
                TimeTrackingItem replaceAllWith = previousItem.withPendingEnd();
                persister.persist(replaceAllWith);
                publisher.publish(new ItemInserted(replaceAllWith));
                return;
            }
        }
        removeActivity(command);
    }

    private boolean previousAndNextActivitiesMatch(TimeTrackingItem previousItem, TimeTrackingItem nextItem) {
        return previousItem.getActivity().equals(nextItem.getActivity());
    }

    @Override
    public void resumeActivity(ResumeActivity command) {
        requireNonNull(command);
        TimeTrackingItem resumedItem = command.itemToResume
                .withPendingEnd()
                .withStart(command.beginningWith);
        persister.persist(resumedItem);
        publisher.publish(new ItemInserted(resumedItem));
    }

    @Override
    public void resumeLastActivity(ResumeLastActivity command) {
        requireNonNull(command);

        Optional<TimeTrackingItem> lastTimeTrackingItem = queries.getLastItem();
        if (lastTimeTrackingItem.isPresent() && lastTimeTrackingItem.get().getEnd().isPresent()) {
            lastTimeTrackingItem.ifPresent(timeTrackingItem -> {
                TimeTrackingItem resumedItem = timeTrackingItem.withPendingEnd().withStart(command.resumeAt);
                persister.persist(resumedItem);
                publisher.publish(new ItemInserted(resumedItem));
            });
        }
    }

    @Override
    public void bulkChangeActivity(Collection<TimeTrackingItem> itemsToChange, String activity) {
        requireNonNull(itemsToChange);
        requireNonNull(activity);
        Collection<ItemPersister.UpdatedItem> updatedItems = persister.updateActivitities(itemsToChange, activity);
        updatedItems.stream()
                .map(updatedItem -> new ItemReplaced(updatedItem.original, updatedItem.updated))
                .forEach(publisher::publish);
    }

    private static class DoNotPublish implements PubSubSupport<Object> {
        @Override
        public void subscribe(Object listener) {
            throw new IllegalStateException();
        }

        @Override
        public boolean unsubscribe(Object listener) {
            throw new IllegalStateException();
        }

        @Override
        public IMessagePublication publish(Object message) {
            return null;
        }

        @Override
        public BusRuntime getRuntime() {
            throw new IllegalStateException();
        }
    }
}

File: src/main/kotlin/org/stt/command/Activities.kt
package org.stt.command

import net.engio.mbassy.bus.BusRuntime
import net.engio.mbassy.bus.IMessagePublication
import net.engio.mbassy.bus.MBassador
import net.engio.mbassy.bus.common.PubSubSupport
import org.stt.model.ItemDeleted
import org.stt.model.ItemInserted
import org.stt.model.ItemReplaced
import org.stt.model.TimeTrackingItem
import org.stt.persistence.ItemPersister
import org.stt.query.Criteria
import org.stt.query.TimeTrackingItemQueries
import org.stt.time.DateTimes
import java.util.*
import java.util.Objects.requireNonNull
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Command "side" of tracking activities
 */
@Singleton
class Activities @Inject
constructor(private val persister: ItemPersister,
            private val queries: TimeTrackingItemQueries,
            publisher: Optional<MBassador<Any>>) : CommandHandler {

    private val publisher: PubSubSupport<Any> = publisher.map { it as PubSubSupport<Any> }.orElseGet { DoNotPublish() }

    override fun addNewActivity(command: NewActivity) {
        requireNonNull(command)
        val newItem = command.newItem
        var potentialItemToReplace = ongoingItemThatWouldEnd(newItem)
        if (potentialItemToReplace == null) {
            potentialItemToReplace = itemWithEditedActivity(newItem)
        }

        if (potentialItemToReplace != null) {
            persister.replace(potentialItemToReplace, command.newItem)
            publisher.publish(ItemReplaced(potentialItemToReplace, command.newItem))
        } else {
            persister.persist(command.newItem)
            publisher.publish(ItemInserted(command.newItem))
        }
    }

    private fun itemWithEditedActivity(newItem: TimeTrackingItem): TimeTrackingItem? {
        val criteria = Criteria()
                .withStartsAt(newItem.start)
        newItem.end?.let { criteria.withEndsAt(it) }
        return queries.queryItems(criteria).findAny().orElse(null)
    }

    private fun ongoingItemThatWouldEnd(newItem: TimeTrackingItem): TimeTrackingItem? {
        return queries.lastItem?.let {
            if (it.end == null && newItem.start == it.start) it else null
        }
    }

    override fun endCurrentActivity(command: EndCurrentItem) {
        requireNonNull(command)
        queries.ongoingItem?.let {
            val derivedItem = it.withEnd(command.endAt)
            persister.replace(it, derivedItem)
            publisher.publish(ItemReplaced(it, derivedItem))
        }
    }

    override fun removeActivity(command: RemoveActivity) {
        requireNonNull(command)
        persister.delete(command.itemToDelete)
        publisher.publish(ItemDeleted(command.itemToDelete))
    }

    override fun removeActivityAndCloseGap(command: RemoveActivity) {
        requireNonNull(command)
        val adjacentItems = queries.getAdjacentItems(command.itemToDelete)
        val previous = adjacentItems.previousItem
        val next = adjacentItems.nextItem

        if (previous != null) {
            if (next != null && previousAndNextActivitiesMatch(previous, next)) {
                val replaceAllWith = next.end?.let { previous.withEnd(it) } ?: previous.withPendingEnd()
                persister.persist(replaceAllWith)
                publisher.publish(ItemInserted(replaceAllWith))
                return
            }
            if (DateTimes.isOnSameDay(previous.start, command.itemToDelete.start) && command.itemToDelete.end == null) {
                val replaceAllWith = previous.withPendingEnd()
                persister.persist(replaceAllWith)
                publisher.publish(ItemInserted(replaceAllWith))
                return
            }
        }
        removeActivity(command)
    }

    private fun previousAndNextActivitiesMatch(previousItem: TimeTrackingItem, nextItem: TimeTrackingItem): Boolean {
        return previousItem.activity == nextItem.activity
    }

    override fun resumeActivity(command: ResumeActivity) {
        requireNonNull(command)
        val resumedItem = command.itemToResume
                .withPendingEnd()
                .withStart(command.beginningWith)
        persister.persist(resumedItem)
        publisher.publish(ItemInserted(resumedItem))
    }

    override fun resumeLastActivity(command: ResumeLastActivity) {
        requireNonNull(command)

        val lastTimeTrackingItem = queries.lastItem
        if (lastTimeTrackingItem?.end != null) {
            val resumedItem = lastTimeTrackingItem.withPendingEnd().withStart(command.resumeAt)
            persister.persist(resumedItem)
            publisher.publish(ItemInserted(resumedItem))
        }
    }

    override fun bulkChangeActivity(itemsToChange: Collection<TimeTrackingItem>, activity: String) {
        requireNonNull(itemsToChange)
        requireNonNull(activity)
        val updatedItems = persister.updateActivitities(itemsToChange, activity)
        updatedItems.stream()
                .map { updatedItem -> ItemReplaced(updatedItem.original, updatedItem.updated) }
                .forEach { publisher.publish(it) }
    }

    private class DoNotPublish : PubSubSupport<Any> {
        override fun subscribe(listener: Any) {
            throw IllegalStateException()
        }

        override fun unsubscribe(listener: Any): Boolean {
            throw IllegalStateException()
        }

        override fun publish(message: Any): IMessagePublication? {
            return null
        }

        override fun getRuntime(): BusRuntime {
            throw IllegalStateException()
        }
    }
}